flasc/analysis/expected_power_analysis.py

```diff
@@ -0,0 +1,779 @@
+            Defaults to ["wd_bin", "ws_bin"].
+        weight_by (str): The method to weight the bins. Defaults to "min".
+        df_freq_pl (pl.DataFrame): A polars dataframe with the frequency of each bin.
+            Defaults to None.
+        percentiles (List[float]): The percentiles to for confidence intervals.
+            Defaults to [2.5, 97.5].
+        remove_any_null_turbine_bins (bool): When computing farm power, remove any bins where
+            and of the test turbines is null.  Defaults to False.
+        cov_terms (str): Use directly computed covariance terms, or fill with zeros or variances.
+            Can be "zero", "var" or "cov".  If "zero" all covariance terms are set to zero.  if
+            "var" all covariance terms are set to the product of the variances.  If "cov" the
+            covariance terms are used as is with missing terms set to product of the
+            variances.  Defaults to "zero".
+
+
+    Returns:
+        Dict[str, Dict[str, float]]: A dictionary containing the uplift results with standard error.
+    """
+    # Get the bin cols without df_name
+    bin_cols_without_df_name = [c for c in bin_cols_in if c != "df_name"]
+    bin_cols_with_df_name = bin_cols_in + ["df_name"]
+
+    # Add the wd and ws bins
+    df_ = _add_wd_ws_bins(
+        df_,
+        wd_cols=wd_cols,
+        ws_cols=ws_cols,
+        wd_step=wd_step,
+        wd_min=wd_min,
+        wd_max=wd_max,
+        ws_step=ws_step,
+        ws_min=ws_min,
+        ws_max=ws_max,
+    )
+
+    # with pl.Config(tbl_cols=-1):
+    #     print(df_)
+
+    # Compute the covariance frame
+    df_cov = _compute_covariance(
+        df_, test_cols=test_cols, bin_cols_with_df_name=bin_cols_with_df_name
+    )
+
+    with pl.Config(tbl_cols=-1):
+        print(df_cov)
+
+    # In current version of code, covarariances are either set to 0 or set to
+    # product of variances
+    if cov_terms == "zero":
+        df_cov = _set_cov_to_zero(df_cov, test_cols=test_cols)
+    elif cov_terms == "var":
+        df_cov = _fill_cov_with_var(df_cov, test_cols=test_cols, fill_all=True)
+    elif cov_terms == "cov":
+        df_cov = _fill_cov_with_var(df_cov, test_cols=test_cols, fill_all=False)
+    else:
+        raise ValueError(f"cov_terms must be 'zero', 'var' or 'cov', not {cov_terms}")
+
+    # with pl.Config(tbl_cols=-1):
+    #     print(df_cov)
+
+    # # If filling missing covariance terms, do it now
+    # if fill_cov_with_var:
+    #     df_cov = _fill_cov_with_var(df_cov, test_cols=test_cols)
+
+    # # If only using the variance, zero out the covariance terms
+    # if variance_only:
+    #     df_cov = _set_cov_to_zero(df_cov, test_cols=test_cols)
+
+    # Apply Null values to covariance and sync across uplift pairs
+    df_cov = _null_and_sync_covariance(
+        df_cov=df_cov,
+        test_cols=test_cols,
+        uplift_pairs=uplift_pairs,
+        bin_cols_without_df_name=bin_cols_without_df_name,
+    )
```

Comment:
As discussed, let's remove the _null_and_sync_covariance function. Most of the operations in the function aren't needed since the output of _compute_covariance will already have Nulls in the right places (and we can add a new test to make sure). The synchronization of Null covariance terms between uplift pair dataframes can instead be accomplished by adding all combinations of cov_t1_t2 to the sync_cols list when calling _synchronize_nulls later.

Reply:
Agreed removed